juju/juju-controller
# Loki support for k8s

This runs on a modified version of Juju 3.0, in which Pebble runs inside the `api-server` container. (@tlm is working on adding this to Juju). Pebble is used to run Promtail inside the `api-server` container, which should send Juju controller logs to Loki.

I've managed to get the modified `juju-controller` charm to deploy successfully and relate to `loki-k8s`. However, the logs are not showing up in Grafana. I need to do some introspection with Pebble to work out what is going on, but I don't know how to use the Pebble API.

## QA steps

Assuming Juju 3.0 with LXD and MicroK8s set up.

Deploy COS Lite on MicroK8s:
```console
$ juju bootstrap microk8s cos
$ juju add-model cos
$ juju deploy cos-lite --channel=edge --trust
$ juju offer loki:logging
Application "loki" endpoints [logging] available at "admin/cos.loki"
```

Bootstrap a k8s controller with this controller charm:
```console
$ juju bootstrap microk8s c --controller-charm-path=ch:barrettj12-cc # modified controller charm
$ juju switch controller
$ juju relate controller cos:cos.loki
```

Now the Juju controller should be connected to Loki. We will verify by using the Grafana dashboard.
```console
$ juju switch cos
$ juju status
Unit             Workload  Agent  Address      Ports  Message         
grafana/0*       active    idle   [ip-address]
$ juju run grafana/0 get-admin-password
admin-password: .....
```

Go to `[ip-address]:3000` in your browser, and log into Grafana with username `admin` and the previously obtained password. On the sidebar, go to "Explore" (the compass), and at the top of the page, change the source to Loki. You should be able to see controller logs using the "log browser".
